go/libraries/doltcore/sqle/read_replica_database.go

```diff
@@ -67,88 +72,175 @@ func NewReadReplicaDatabase(ctx context.Context, db Database, remoteName string,
 		return EmptyReadReplica, err
 	}
 
-	headRef := rsr.CWBHeadRef()
-	refSpecs, err := env.GetRefSpecs(rsr, remoteName)
-
-	var remoteTrackRef ref.DoltRef
-	var foundRef bool
-	for _, refSpec := range refSpecs {
-		trackRef := refSpec.DestRef(headRef)
-		if trackRef != nil {
-			remoteTrackRef = trackRef
-			foundRef = true
-			break
-		}
-	}
-	if !foundRef {
-		return EmptyReadReplica, env.ErrInvalidRefSpecRemote
+	branches, err := db.ddb.GetBranches(ctx)
+	localBranches := make(map[string]ref.DoltRef, len(branches))
+	for _, br := range branches {
+		localBranches[br.String()] = br
 	}
 
 	return ReadReplicaDatabase{
-		Database:       db,
-		headRef:        headRef,
-		remoteTrackRef: remoteTrackRef,
-		remote:         remote,
-		tmpDir:         tmpDir,
-		srcDB:          srcDB,
+		Database:      db,
+		remote:        remote,
+		tmpDir:        tmpDir,
+		srcDB:         srcDB,
+		localBranches: localBranches,
+		headRef:       rsr.CWBHeadRef(),
 	}, nil
 }
 
 func (rrd ReadReplicaDatabase) StartTransaction(ctx *sql.Context, tCharacteristic sql.TransactionCharacteristic) (sql.Transaction, error) {
 	if rrd.srcDB != nil {
-		err := rrd.pullFromReplica(ctx)
+		err := rrd.PullFromReplica(ctx)
 		if err != nil {
-			return nil, err
+			err = fmt.Errorf("replication failed: %w", err)
+			if !SkipReplicationWarnings() {
+				return nil, err
+			}
+			ctx.GetLogger().Warn(err.Error())
 		}
 	} else {
-		ctx.GetLogger().Warn("replication failure; dolt_replication_remote value is misconfigured")
+		ctx.GetLogger().Warn("replication failed; dolt_replication_remote value is misconfigured")
 	}
 	return rrd.Database.StartTransaction(ctx, tCharacteristic)
 }
 
-func (rrd ReadReplicaDatabase) pullFromReplica(ctx *sql.Context) error {
-	ddb := rrd.Database.DbData().Ddb
+func (rrd ReadReplicaDatabase) PullFromReplica(ctx context.Context) error {
+	_, headsArg, ok := sql.SystemVariables.GetGlobal(ReplicateHeadsKey)
+	if !ok {
+		return sql.ErrUnknownSystemVariable.New(ReplicateHeadsKey)
+	}
+
+	_, allHeads, ok := sql.SystemVariables.GetGlobal(ReplicateAllHeadsKey)
+	if !ok {
+		return sql.ErrUnknownSystemVariable.New(ReplicateAllHeadsKey)
+	}
+
+	switch {
+	case headsArg != "" && allHeads == int8(1):
+		return fmt.Errorf("%w; cannot set both 'dolt_replicate_heads' and 'dolt_replicate_all_heads'", ErrInvalidReplicateHeadSetting)
+	case headsArg != "":
+		heads, ok := headsArg.(string)
+		if !ok {
+			return sql.ErrInvalidSystemVariableValue.New(heads)
+		}
+		branches := parseBranches(heads)
+		err := pullBranches(ctx, rrd, branches)
+		if err != nil {
+			return err
+		}
+	case allHeads == int8(1):
+		err := rrd.srcDB.Rebase(ctx)
+		if err != nil {
+			return err
+		}
+
+		refs, err := rrd.srcDB.GetBranches(ctx)
```

Comment:
This doesn't work to get a new branch that was pushed, right? Unless you did a fetch to include branches on the remote?